File: python/2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

#%%
from typing import Optional

#%%
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

#%%
def list_to_ListNode(l):
    head = None
    cur = None
    for i in range(len(l)):
        node = ListNode(l[i])

        if not head:
            head = node

        if not cur:
            cur = node
        else:
            cur.next = node
            cur = node
    
    return head

def ListNode_to_list(listnode):
    l = []
    cur = listnode
    while cur:
        l.append(cur.val)
        cur = cur.next
    return l

#%%
# @lc code=start
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        answer = ListNode()
        tail = answer
        carry = 0

        while l1 or l2 or carry:
            num1 = l1.val if l1 else 0
            num2 = l2.val if l2 else 0

            carry, out = divmod(num1 + num2 + carry, 10)

            tail.next = ListNode(out)
            tail = tail.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return answer.next
# @lc code=end

#%%
l1 = list_to_ListNode([2, 4, 3])
l2 = list_to_ListNode([5, 6, 4])
answer = ListNode_to_list(Solution().addTwoNumbers(l1, l2))

print(answer)

#%%

File: csharp/2.add-two-numbers.cs
/*
 * @lc app=leetcode id=2 lang=csharp
 *
 * [2] Add Two Numbers
 */
using System.Collections;


var l1 = ListNode.FromEnumerable(new[] { 2, 4, 3 });
var l2 = ListNode.FromEnumerable(new[] { 5, 6, 4 });
Console.WriteLine( new Solution().AddTwoNumbers(l1, l2).ToString() );


// Definition for singly-linked list.
public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int val=0, ListNode next=null) {
        this.val = val;
        this.next = next;
    }

    public static ListNode FromEnumerable(IEnumerable<int> enumerable) {
        var head = new ListNode();
        var cur = head;
        foreach(var item in enumerable){
            cur.next = new ListNode() { val = item };
            cur = cur.next;
        }
        return head.next;
    }

    public IEnumerable<int> ToEnumerable() {
        var cur = this;
        do {
            yield return cur.val;
            cur = cur.next;
        } while(cur != null); 
    }

    public override string ToString()
    {
        var l = this.ToEnumerable();
        return $"[{string.Join(", ", l)}]"; 
    }
}

// @lc code=start
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        var head = new ListNode();
        var cur = head;
        int carry = 0;

        while(l1 != null || l2 != null || carry != 0){
            int sum = (l1?.val??0) + (l2?.val??0) + carry;

            cur.next = new ListNode(){ val = sum%10 };
            cur = cur.next;

            carry = sum/10;
            l1 = l1?.next;
            l2 = l2?.next;
        }

        return head.next;
    }
}
// @lc code=end

